site/source/_posts/2019/01/28/Version-5-Dojo.md

```diff
@@ -0,0 +1,234 @@
+
+```json
+"build-app": {
+	"build-time-render": {
+		"root": "app",
+		"paths": [
+			"home",
+			"about",
+		]
+	}
+}
+```
+
+### Introducing Dojo Blocks
+
+Dojo Blocks is a new mechanism leveraging Build Time Rendering that allows executing code in node.js as part of the build. The results from the execution are written to a cache that can then be transparently used in the same way at runtime in the browser. This opens up new opportunities for performing operations that might be not possible or unperformant in a browser.
+
+For example, a Dojo Block module could read a group of markdown files, transform them into VNodes, and make them available to render in the application, all at build time. The result of this Dojo Block module is then cached into the application bundle for use at runtime in the browser.
+
+A Dojo Block module gets used like any other [meta][MetaReadme] in any Dojo widget. As a result there is no extensive configuration or alternative authoring patterns needed.
+
+For example, a block module could read a text file and return the content to the application:
+
+```ts
+// src/read-file.block.ts
+import * as fs from 'fs';
+import { resolve } from 'path';
+
+export default (path: string) => {
+	path = resolve(__dirname, path);
+	return fs.readFileSync(path, 'utf8');
+}
+```
+
+Using the `text.block.ts` module in your application with the [`Block` meta](https://github.com/dojo/framework/blob/master/src/widget-core/meta/Block.ts):
+
+```ts
+// src/MyBlockWidget.ts
+import Block from '@dojo/framework/widget-core/meta/Block';
+import WidgetBase from '@dojo/framework/widget-core/WidgetBase';
+import { v } from '@dojo/framework/widget-core/d';
+
+import readFile from './read-file.block';
+
+export default class MyBlockWidget extends WidgetBase {
+	protected render() {
+		const message = this.meta(Block).run(readFile)('../content/hello-dojo-blocks.txt');
+		return v('div', [ message ]);
+	}
+}
+```
+
+This widget runs the `read-file.block.ts` module at build time, loading the file path passed which gets used as the children in the widget. The rendered application's HTML and CSS get written to the built `index.html`:
+
+```html
+<html>
+<head>
+	<title>Dojo Blocks</title>
+	<meta name="theme-color" content="#222127">
+	<meta name="viewport" content="width=device-width, initial-scale=1">
+</head>
+<body>
+	<div id="root"><div>Hello Dojo Block World!</div></div>
+	<script type="text/javascript" src="bootstrap.js"></script></body>
+</html>
+```
+
+### Simplifying Testing with Assertion Templates
+
+Assertion Templates make testing widgets with the testing harness even easier. Instead of having to manually curate each expectedRender result per test, you can now use the Assertion Template primitive to easily modify and layer outputs for the expected render.
```

Comment:
```suggestion
Assertion Templates make testing widgets with the testing harness even easier. Instead of needing to manually curate each `expectedRender` result per test, you can now use the Assertion Template primitive to easily modify and layer outputs for the expected render.
```